source/Octopus.Tentacle/Kubernetes/KubernetesPodMonitor.cs

```diff
@@ -0,0 +1,166 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using k8s;
+using k8s.Models;
+using Octopus.Diagnostics;
+using Octopus.Tentacle.Contracts;
+
+namespace Octopus.Tentacle.Kubernetes
+{
+    public interface IKubernetesPodMonitor
+    {
+        Task StartAsync(CancellationToken token);
+    }
+
+    public interface IKubernetesPodStatusProvider
+    {
+        PodStatus? TryGetPodStatus(ScriptTicket scriptTicket);
+    }
+
+    public class KubernetesPodMonitor : IKubernetesPodMonitor, IKubernetesPodStatusProvider
+    {
+        readonly IKubernetesPodService podService;
+        readonly ISystemLog log;
+        readonly Dictionary<ScriptTicket, PodStatus> podStatusLookup = new();
+
+        public KubernetesPodMonitor(IKubernetesPodService podService, ISystemLog log)
+        {
+            this.podService = podService;
+            this.log = log;
+        }
+
+        async Task IKubernetesPodMonitor.StartAsync(CancellationToken cancellationToken)
+        {
+            while (!cancellationToken.IsCancellationRequested)
+            {
+                //initially load all the pods and their status's
+                var initialResourceVersion = await InitialLoadAsync(cancellationToken);
+
+                // We start the watch from the resource version we initially loaded.
+                // This means we only receive events that occur after the resource version
+                await podService.WatchAllPods(initialResourceVersion, async (type, pod) =>
+                    {
+                        await Task.CompletedTask;
+
+                        try
+                        {
+                            log.Verbose($"Received {type} event for pod {pod.Name()}");
+
+                            var scriptTicket = pod.GetScriptTicket();
+
+                            switch (type)
```

Comment:
Can we extract this logic and unit test?

Reply:
yep

Reply:
@flin-8 Added an initial set of tests in 03baaa4

Reply:
Good start 👍 

I usually prefer to extract out classes instead of making methods internal/public just for tests (we could potentially separate out a class for the state transition from the watching) - but don't let that stop you from merging

Reply:
ok, I'll keep that in mind. This area is getting a bit of rework with the logs monitoring, so I'll try and clean it up there 👍